hutch_python/namespace.py

```diff
@@ -131,28 +132,58 @@ def tree_namespace(scope=None):
     namespace: `IterableNamespace`
     """
     logger.debug('Create tree_namespace scope=%s', scope)
-    tree_space = IterableNamespace()
     scope_objs = extract_objs(scope=scope, stack_offset=1)
 
-    for name, obj in scope_objs.items():
-        logger.debug('Add %s to tree namespace', name)
-        upper_space = tree_space
-        keys = name.split('_')[:-1]
+    def nested_defaultdict():
+        return defaultdict(nested_defaultdict)
+
+    tree_dict = nested_defaultdict()
 
-        if keys:
-            # Add key to existing namespace branch, create new if needed
+    # Decide which nodes are objs, which are namespaces
+    for name, obj in scope_objs.items():
+        curr_dict = tree_dict
+        keys = name.split('_')
+        valid_keys = True
+        for key in keys:
+            if iskeyword(key):
+                logger.warning('Issue in tree_namespace, %s in %s is a '
+                               'reserved python keyword, omitting from tree',
+                               key, name)
+                valid_keys = False
+                break
+            if not key.isidentifier():
+                logger.warning('Issue in tree_namespace, %s in %s is an '
+                               'invalid python identifier, omitting from tree',
+                               key, name)
+                valid_keys = False
+                break
+        if valid_keys:
             for key in keys:
-                name = strip_prefix(name, key)
-                # Force lowercase
-                key = key.lower()
-                if not hasattr(upper_space, key):
-                    setattr(upper_space, key, IterableNamespace())
-                upper_space = getattr(upper_space, key)
-            if hasattr(upper_space, name):
-                logger.warning(('Tried to add {} to {}, but something was '
-                                'already there. Two devices share the same '
-                                'name!'.format(name, upper_space)))
+                curr_dict = curr_dict[key]
+            curr_dict['_obj'] = obj
+
+    # Unpack tree_dict to make tree_space
+    def unpack_node(node):
+        obj = node.pop('_obj', None)
+        if obj is None:
+            obj = IterableNamespace()
+        for key, value in node.items():
+            if hasattr(obj, key):
+                # This means we would be overwriting something important
+                logger.warning('Issue in tree_namespace, tried to overrwrite'
+                               '%s.%s=%s', obj, key, getattr(obj, key))
             else:
-                setattr(upper_space, name, obj)
+                sub_node = unpack_node(value)
+                try:
+                    setattr(obj, key, sub_node)
```

Comment:
I'm kind of against tacking things on here. What if it overwrites an existing attribute? 

```python
motor = EpicsMotor(...)

def motor_timeout(...):
    pass
```

Reply:
I would only do this if it is an `IterableNamespace`

Reply:
I check if it is an existing attribute a few lines up.